state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+            return Err(Error::UnknownEventSubscription(
+                subscription_id,
+                format!("{:?}", old_subscription),
+            ));
+        }
+
+        // Update the event key subscriptions to include the new subscription
+        for event_key in event_keys {
+            self.event_key_subscriptions
+                .entry(event_key)
+                .and_modify(|subscriptions| {
+                    subscriptions.insert(subscription_id);
+                })
+                .or_insert_with(|| HashSet::from_iter(vec![subscription_id].iter().cloned()));
+        }
+
+        Ok(EventNotificationListener {
+            notification_receiver,
+        })
+    }
+
+    /// Returns a ReconfigNotificationListener that can be monitored for
+    /// reconfiguration events. Subscribers will be sent a notification
+    /// containing all new on-chain configuration values whenever a new epoch
+    /// begins. Note: if the notification buffer fills up too quickly, older
+    /// notifications will be dropped. As such, it is the responsibility of the
+    /// subscriber to ensure notifications are processed in a timely manner.
+    pub fn subscribe_to_reconfigurations(&mut self) -> Result<ReconfigNotificationListener, Error> {
+        let (notification_sender, notification_receiver) =
+            diem_channel::new(QueueStyle::KLAST, RECONFIG_NOTIFICATION_CHANNEL_SIZE, None);
+
+        // Create a new reconfiguration subscription
+        let subscription_id = self.get_new_subscription_id();
+        let reconfig_subscription = ReconfigSubscription {
+            subscription_id,
+            notification_sender,
+        };
+
+        // Store the new subscription
+        if let Some(old_subscription) = self
+            .reconfig_subscriptions
+            .insert(subscription_id, reconfig_subscription)
+        {
+            return Err(Error::UnknownReconfigSubscription(
+                subscription_id,
+                format!("{:?}", old_subscription),
+            ));
+        }
+
+        Ok(ReconfigNotificationListener {
+            notification_receiver,
+        })
+    }
+
+    fn get_new_subscription_id(&mut self) -> u64 {
+        self.next_subscription_id.fetch_add(1, Ordering::Relaxed)
+    }
+
+    /// This notifies all the event subscribers of the new events found at the
+    /// specified version. If a reconfiguration event (i.e., new epoch) is found,
+    /// this method will return true.
+    fn notify_event_subscribers(
+        &mut self,
+        version: Version,
+        events: Vec<ContractEvent>,
+    ) -> Result<bool, Error> {
+        let mut reconfig_event_found = false;
+        let mut event_subscription_ids_to_notify = HashSet::new();
+
+        for event in events.iter() {
+            let event_key = *event.key();
+
+            // Process all subscriptions for the current event
+            if let Entry::Occupied(subscription_ids) = self.event_key_subscriptions.entry(event_key)
```

Comment:
this looks weird, why not use if let Some(..) = self.map.get(&event_key)?

Reply:
Beautiful!